### ISSUE TITLE
FP to int conversion relies on UB

### ISSUE THREAD

**Issue by jfbastien** [WebAssembly/binaryen#420]
Looking into #404 more I get the following:
```
(/home/travis/build/WebAssembly/binaryen/bin/binaryen-shell+0x496f21): runtime error: value 9.22337e+18 is outside the range of representable values of type 'long'
```
The problem is with doing FP to int conversion unconditionally. The "obvious" fix of checking bounds and then doing the conversion doesn't work either because some of the values close to int min / max convert differently based on rounding, and many of them convert to the same integer.

I think the best approach there is to rely on exponent checks, I'll dig further.

What I mean by "obvious" but incorrect:
```
          if (val <= (double)std::numeric_limits<int64_t>::max() &&
              val >= (double)std::numeric_limits<int64_t>::min()) {
            return Literal(int64_t(val));
          }
          trap("i64.truncSFloat overflow");
          return Literal(int64_t(0));
```

**Comment by binji** [comment]
In sexpr-wasm I perform the check using integers instead of floats: https://github.com/WebAssembly/sexpr-wasm-prototype/blob/master/src/wasm-interpreter.c#L83

The table is sorted by the binary representation of the float, so it's a little easier to understand the comparisons.

**Comment by jfbastien** [comment]
@binji ah cool, that's about what I had in mind, but using `numeric_limits`'s `exponent` and `digits`  to avoid hard-coding the exact representation.

Do you mind importing what you have to binaryen? Seems silly to rewrite if you already have it.

**Comment by binji** [comment]
sure, I'll take a look.

**Comment by binji** [comment]
https://github.com/WebAssembly/binaryen/pull/421

**Comment by jfbastien** [comment]
Fixed by #421 (will re-run UBsan bot).
